src/main/java/org/broadinstitute/hellbender/tools/spark/sv/AlignedAssemblyOrExcuse.java

```diff
@@ -111,41 +114,80 @@ public FermiLiteAssembly getAssembly() {
     /**
      * write a SAM file containing records for each aligned contig
      */
-    public static void writeSAMFile( final String samFile, final PipelineOptions pOpts,
-                                     final SAMFileHeader header,
-                                     final List<AlignedAssemblyOrExcuse> alignedAssemblyOrExcuseList ) {
+    static void writeSAMFile( final String samFile, final PipelineOptions pOpts,
+                              final SAMFileHeader header,
+                              final List<AlignedAssemblyOrExcuse> alignedAssemblyOrExcuseList ) {
         try ( final OutputStream os = BucketUtils.createFile(samFile, pOpts) ) {
             final SAMTextWriter writer = new SAMTextWriter(os);
             writer.setSortOrder(SAMFileHeader.SortOrder.queryname, true);
             writer.setHeader(header);
-            final List<String> refNames =
-                    header.getSequenceDictionary().getSequences().stream()
-                            .map(SAMSequenceRecord::getSequenceName).collect(Collectors.toList());
-            for ( final AlignedAssemblyOrExcuse alignedAssemblyOrExcuse : alignedAssemblyOrExcuseList ) {
-                if ( alignedAssemblyOrExcuse.getErrorMessage() != null ) continue;
-                final FermiLiteAssembly assembly = alignedAssemblyOrExcuse.getAssembly();
-                final int assemblyId = alignedAssemblyOrExcuse.getAssemblyId();
-                final List<List<BwaMemAlignment>> allAlignments = alignedAssemblyOrExcuse.getContigAlignments();
-                final int nContigs = assembly.getNContigs();
-                for ( int contigIdx = 0; contigIdx != nContigs; ++contigIdx ) {
-                    final List<BwaMemAlignment> alignments = allAlignments.get(contigIdx);
-                    if ( alignments.isEmpty() ) continue;
-                    final FermiLiteAssembly.Contig contig = assembly.getContig(contigIdx);
-                    final String readName = String.format("asm%06d:tig%05d", assemblyId, contigIdx);
-                    final byte[] calls = contig.getSequence();
-                    for ( final BwaMemAlignment alignment : alignments ) {
-                        writer.addAlignment(BwaMemAlignmentUtils.applyAlignment(readName, calls, null, null, alignment,
-                                refNames, header, false, false));
-                    }
-                }
-            }
+            filterFailedAssemblyAndConvertToSamRecords(header, alignedAssemblyOrExcuseList).forEach(contigs -> contigs.forEach(ls -> ls.forEach(writer::addAlignment)) );
```

Comment:
I was actually using such style that you suggested, but the same code block is used in a different place `StructuralVariationDiscoveryPipelineSpark.InMemoryAlignmentParser.filterAndConvertToAlignedContigViaSAM()`, which was trying to go through the assemblies one by one. For the code to accommodate the needs there, it needs at least a `List<List<SAMRecord>>` where the outer level is for contigs. If everything alignment is to be flattened out to a single list, it would be difficult for that piece of code knowing which SAMRecord comes from which contig.
So I'll go with the 2-D list.